tokio/src/sync/notify.rs

```diff
@@ -0,0 +1,496 @@
+use crate::loom::sync::atomic::AtomicU8;
+use crate::loom::sync::Mutex;
+use crate::util::linked_list::{self, LinkedList};
+
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::atomic::Ordering::SeqCst;
+use std::task::{Context, Poll, Waker};
+
+/// Notify a single task to wake up.
+///
+/// `Notify` provides a basic mechanism to notify a single task of an event.
+/// `Notify` itself does not carry any data. Instead, it is to be used to signal
+/// another task to perform an operation.
+///
+/// The synchronization details of `Notify` are similar to
+/// [`thread::park`][park] and [`Thread::unpark`][unpark] from std. A [`Notify`]
+/// value contains a single permit. [`Notify::recv`] waits for the permit to be
+/// made available, consumes the permit, and resumes. [`Notify::notify_one`]
+/// sets the permit, waking a pending task if there is one.
+///
+/// `Notify` can be thought of as a [`Semaphore`] starting with 0 permits.
+/// [`Notify::recv`] waits for a permit to become available and
+/// [`Notify::notify_one`] sets a permit **if there currently are no available
+/// permits**.
+///
+/// If `notify_one` is called **before** `recv()`, then the next call to
+/// `recv()` will complete immediately, consuming the permit. Any subsequent
+/// calls to `recv()` will wait for a new permit.
+///
+/// If `notify_one` is called **multiple** times before `recv()`, only a
+/// **single** permit is stored. The next call to `recv()` will complete
+/// immediately, but the one after will wait for a new permit.
+///
+/// # Examples
+///
+/// Basic usage.
+///
+/// ```
+/// use tokio::sync::Notify;
+/// use std::sync::Arc;
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let notify = Arc::new(Notify::new());
+///     let notify2 = notify.clone();
+///
+///     tokio::spawn(async move {
+///         notify2.recv().await;
+///         println!("received a notification");
+///     });
+///
+///     notify.notify_one();
```

Comment:
Maybe include a `println!("sent notification")` before here, and say that they _will_ print in order.